plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/GcpEmulationTab.java

```diff
@@ -0,0 +1,237 @@
+
+    Composite projectSelectorComposite = new Composite(composite, SWT.NONE);
+    final Text filterField = new Text(projectSelectorComposite,
+        SWT.BORDER | SWT.SEARCH | SWT.ICON_SEARCH | SWT.ICON_CANCEL);
+
+    projectSelector = new ProjectSelector(projectSelectorComposite);
+    projectSelector.addSelectionChangedListener(new ISelectionChangedListener() {
+      @Override
+      public void selectionChanged(SelectionChangedEvent event) {
+        if (!initializingUiValues) {
+          boolean emptySelection = projectSelector.getSelectProjectId().isEmpty();
+          boolean savedIdAvailable = projectSelector.isProjectIdAvailable(gcpProjectIdModel);
+          if (!emptySelection || savedIdAvailable) {
+            gcpProjectIdModel = projectSelector.getSelectProjectId();
+            updateLaunchConfigurationDialog();
+          }
+        }
+      }
+    });
+
+    filterField.setMessage(Messages.getString("project.filter.hint")); //$NON-NLS-1$
+    filterField.addModifyListener(new ModifyListener() {
+      @Override
+      public void modifyText(ModifyEvent event) {
+        projectSelector.setFilter(filterField.getText());
+      }
+    });
+
+    // Service key row
+    new Label(composite, SWT.LEAD).setText(Messages.getString("label.service.key")); //$NON-NLS-1$
+    serviceKeyInput = new Text(composite, SWT.BORDER);
+    serviceKeyInput.addModifyListener(new ModifyListener() {
+      @Override
+      public void modifyText(ModifyEvent event) {
+        updateLaunchConfigurationDialog();
+      }
+    });
+    Button browse = new Button(composite, SWT.NONE);
+    browse.setText(Messages.getString("button.browse")); //$NON-NLS-1$
+
+    GridDataFactory.fillDefaults().align(SWT.BEGINNING, SWT.TOP).applyTo(projectLabel);
+    GridDataFactory.fillDefaults().span(2, 1).applyTo(accountSelector);
+    GridLayoutFactory.swtDefaults().numColumns(3).generateLayout(composite);
+
+    GridDataFactory.fillDefaults().span(2, 1).applyTo(projectSelectorComposite);
+    GridDataFactory.fillDefaults().applyTo(filterField);
+    GridDataFactory.fillDefaults().grab(true, false).hint(SWT.DEFAULT, 200)
+        .applyTo(projectSelector);
+    GridLayoutFactory.fillDefaults().spacing(0, 0).generateLayout(projectSelectorComposite);
+
+    setControl(composite);
+
+    gcpIcon = SharedImages.GCP_IMAGE_DESCRIPTOR.createImage();
+    composite.addDisposeListener(new ImageDisposer(gcpIcon));
+  }
+
+  @Override
+  public void setDefaults(ILaunchConfigurationWorkingCopy configuration) {
+    // No particular default values to set in a newly created configuration.
+  }
+
+  @Override
+  public void initializeFrom(ILaunchConfiguration configuration) {
+    accountEmailModel = getAttribute(configuration, ATTRIBUTE_ACCOUNT_EMAIL, ""); //$NON-NLS-1$
+    gcpProjectIdModel = getAttribute(configuration, ATTRIBUTE_GCP_PROJECT, ""); //$NON-NLS-1$
+    String serviceKey = getAttribute(configuration, ATTRIBUTE_SERVICE_KEY, ""); //$NON-NLS-1$
```

Comment:
So what happens if the user also adds a `GOOGLE_APPLICATION_CREDENTIALS` or `GOOGLE_CLOUD_PROJECT ` value to the environment tab?  Which wins?

One approach would be for this tab to actually store and retrieve its values from the launch config's environment data.

Reply:
I have a comprehensive plan, which I am already doing for Dataflow — it already has an account selector, whose login info is being passed as `GOOGLE_APPLICATION_CREDENTIALS`. Basically, I will do the following:

https://github.com/GoogleCloudPlatform/google-cloud-eclipse/blob/master/plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/launcher/DataflowPipelineLaunchDelegate.java#L183

Reply:
What are you going to do for standard Java launches?

If you store values in the environment, then I think you're done.

Reply:
> What are you going to do for standard Java launches?

I think we can just do the same thing. We anyway have to extend the standard Java launches to add this tab. I just hope we can also extend the launch so that we can extract values from our tab.

> If you store values in the environment, then I think you're done.

That's true, but I'm afraid programmatically adding environment values there may confuse users and actually costs us more complexity in implementation. I think doing this is much simpler, and I don't see any issue.

Reply:
Oh, you actually meant adding env vars to the launch config attribute, right? I thought you meant adding env vars to the _Environments_ tab UI. I agre that is what we should do, now knowing that it seems not possible to extend the vanilla-Java launch delegate.

Reply:
> One approach would be for this tab to actually store and retrieve its values from the launch config's environment data.

After some fiddling, I realized this doesn't work nicely, unless we tightly interwire the _Environments_ tab and our tab. For example, suppose the current `GOOGLE_APPLICATION_CREDENTIALS` is `foo.json`. This value will be loaded in the service key field in our tab. The user will also be able to see it in the _Environments_ tab. Now, in the _Environments_ tab, suppose the user modifies this value to `bar.json`. When the user presses _Apply_, what happens is that the value is immediately reverted back to `foo.json`, because the framework will call `performApply()` of our tab as well (even though our tab is invisible). This will overwrite `bar.json` with `foo.json` that is in the service key text field. This problem hit me, as I closed the run config right way not knowing that the value was reverted.

Basically, it seems not a good idea for a tab to make changes to a model of a different tab, unless you are willing to update UIs of the other tab simultaneously. I don't think we want to extend the _Environments_ tab to hack this. It won't work in general cases either.

Reply:
You'll need to go a bit deeper and implement the `activated()` and `deactivated()` APIs.  That way you'll know when to ignore your text field.

Reply:
OK, that sounded like a bit of abusing the tab framework, but looks like we can utilize it to make things work reasonably acceptable. This results in some minute annoyances (confirming save even when values have restored to its original state, and another due to #2571), but I think they are pretty insignificant issues.